### PR TITLE
[3409][IMP] stock_product_shelfinfo: create new model area1 & area2 and change the fields type to many2one on product.shelfinfo model

### PR DESCRIPTION
**Pull request by rinaldifirdaus**
[3409](https://www.quartile.co/web?db=quartile&token=dMRr78CMkEX6gyy9nTiv#id=3409&action=771&model=project.task&view_type=form&menu_id=505)

### CODE REVIEW

**Comment by rinaldifirdaus** [comment]
In this update, i added area 1 & area 2 menu items and actions to inventory/configurations menu.
currently still got an error when input the area 1 & area 2 on Product Shelf Information menu as follows:

```
Traceback (most recent call last):
  File "/home/rinaldi/ws/axls/odoo/odoo/api.py", line 983, in get
    cache_value = field_cache[record._ids[0]]
KeyError: <NewId 0x7fcee585b130>

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/rinaldi/ws/axls/odoo/odoo/fields.py", line 1160, in __get__
    value = env.cache.get(record, self)
  File "/home/rinaldi/ws/axls/odoo/odoo/api.py", line 990, in get
    raise CacheMiss(record, field)
odoo.exceptions.CacheMiss: 'product.shelfinfo(<NewId 0x7fcee585b130>,).name'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/rinaldi/ws/axls/odoo/odoo/http.py", line 1583, in _serve_db
    return service_model.retrying(self._serve_ir_http, self.env)
  File "/home/rinaldi/ws/axls/odoo/odoo/service/model.py", line 134, in retrying
    result = func()
  File "/home/rinaldi/ws/axls/odoo/odoo/http.py", line 1610, in _serve_ir_http
    response = self.dispatcher.dispatch(rule.endpoint, args)
  File "/home/rinaldi/ws/axls/odoo/odoo/http.py", line 1807, in dispatch
    result = self.request.registry['ir.http']._dispatch(endpoint)
  File "/home/rinaldi/ws/axls/odoo/odoo/addons/base/models/ir_http.py", line 154, in _dispatch
    result = endpoint(**request.params)
  File "/home/rinaldi/ws/axls/odoo/odoo/http.py", line 696, in route_wrapper
    result = endpoint(self, *args, **params_ok)
  File "/home/rinaldi/ws/axls/odoo/addons/web/controllers/dataset.py", line 42, in call_kw
    return self._call_kw(model, method, args, kwargs)
  File "/home/rinaldi/ws/axls/odoo/addons/web/controllers/dataset.py", line 33, in _call_kw
    return call_kw(request.env[model], method, args, kwargs)
  File "/home/rinaldi/ws/axls/odoo/odoo/api.py", line 461, in call_kw
    result = _call_kw_multi(method, model, args, kwargs)
  File "/home/rinaldi/ws/axls/odoo/odoo/api.py", line 448, in _call_kw_multi
    result = method(recs, *args, **kwargs)
  File "/home/rinaldi/ws/axls/odoo/odoo/models.py", line 6513, in onchange
    todo = [
  File "/home/rinaldi/ws/axls/odoo/odoo/models.py", line 6516, in <listcomp>
    if name not in done and snapshot0.has_changed(name)
  File "/home/rinaldi/ws/axls/odoo/odoo/models.py", line 6299, in has_changed
    return self[name] != record[name]
  File "/home/rinaldi/ws/axls/odoo/odoo/models.py", line 5858, in __getitem__
    return self._fields[key].__get__(self, type(self))
  File "/home/rinaldi/ws/axls/odoo/odoo/fields.py", line 1209, in __get__
    self.compute_value(recs)
  File "/home/rinaldi/ws/axls/odoo/odoo/fields.py", line 1368, in compute_value
    records._compute_field_value(self)
  File "/home/rinaldi/ws/axls/odoo/odoo/models.py", line 4193, in _compute_field_value
    fields.determine(field.compute, self)
  File "/home/rinaldi/ws/axls/odoo/odoo/fields.py", line 97, in determine
    return needle(*args)
  File "/home/rinaldi/ws/axls/axls-custom/stock_product_shelfinfo/models/product_shelfinfo.py", line 70, in _compute_name
    record.name += record.area1
TypeError: can only concatenate str (not "product.shelfinfo_area1") to str

The above server error caused the following client error:
RPC_ERROR: Odoo Server Error
    at makeErrorFromResponse (https://07.dev2qrtl.com/web/assets/1462-95cae11/web.assets_backend.min.js:970:163)
    at XMLHttpRequest.<anonymous> (https://07.dev2qrtl.com/web/assets/1462-95cae11/web.assets_backend.min.js:978:13)
```
need to check the _compute_name functions..
will continue tomorrow (05/05).

File: stock_product_shelfinfo/models/product_shelfinfo.py [hunk 1]
```diff
@@ -68,3 +68,29 @@ def _compute_name(self):
...
+    _name = "product.shelfinfo_area1"
+    _description = "Product Shelfinfo Area 1"
+
+    area1 = fields.Char("Area 1")
+
+    def name_get(self):
+        result = []
+        for rec in self:
+            result.append((rec.id, rec.area1))
+        return result
```
Comment (yostashiro) [inline stock_product_shelfinfo/models/product_shelfinfo.py]:
```suggestion
    name = fields.Char(required=True)
```

File: stock_product_shelfinfo/models/product_shelfinfo.py [hunk 2]
```diff
@@ -68,3 +68,29 @@ def _compute_name(self):
...
+    _name = "product.shelfinfo_area2"
+    _description = "Product Shelfinfo Area 2"
+
+    area2 = fields.Char("Area 2")
+
+    def name_get(self):
+        result = []
+        for rec in self:
+            result.append((rec.id, rec.area1))
+        return result
```
Comment (yostashiro) [inline stock_product_shelfinfo/models/product_shelfinfo.py]:
```suggestion
    name = fields.Char(required=True)
``

File: stock_product_shelfinfo/models/product_shelfinfo.py
Comment (yostashiro) [inline stock_product_shelfinfo/models/product_shelfinfo.py:1]: Can we combine the models (Area1 and Area2) into one, and have an indicator (many2many tag) to define which field(s) (area1 and/or area2) the record can be assigned to?

Please split the file in a manner to keep one model per file.

**Review by yostashiro — CHANGES_REQUESTED** [review]

**Comment by rinaldifirdaus** [comment]
@yostashiro 
In the latest commit, the error mentioned on previous commit has been resolved by make an adjustment on _compute_name function.
the models of area 1 & area 2 has been combined to be product_shelfinfo_area model.

> have an indicator (many2many tag) to define which field(s) (area1 and/or area2) the record can be assigned to?
I couldn't find the idea on how to implement the many2many tag on this module. as another idea, i create a selection field "type" along with each name field, so that users can define the type of each area name. then on product shelf info menu i call each record on each fields using domain of the type field.

please kindly your advise. Thanks!

**Comment by yostashiro** [comment]
> the models of area 1 & area 2 has been combined to be product_shelfinfo_area model.

@rinaldifirdaus I have reverted the strategy back to separating models for areas and position after seeing the actual data in the production database. Combining models might add some flexibility for the future, but it comes with the cost of degraded UX, and I judged that the cost overwighs the benefit.

**Review by yostashiro — APPROVED** [review]
@rinaldifirdaus Upgrading the module will wipe out the existing area and location data in shelf info.  Please take care of the data migration from existing fields to the new ones (i.e. prepare import files beforehand).

**Pull request merged**